### PR TITLE
Add types

### PR DESCRIPTION
**Pull request by ocavue**
Some questions or TODOs:

- [x] This PR is pending the release of https://github.com/micromark/micromark/pull/25
- [x] Is there any difference between micromark's `Token` and mdast-util-from-markdown's `Token`?
- [x] Is it okay to just import `Buffer` from `micromark`? Or do we need to re-define our own interface for Node's Buffer？

### CODE REVIEW

**Comment by wooorm** [comment]
2. from-markdown returns a root. Types for mdast are available in [`@types/mdast`](http://npmjs.com/@types/mdast)
3. As this project builds on the parts that micromark’s parse halve also takes, I believe it’s fine to use the types defined there here too!

An example of a typical extension is https://github.com/syntax-tree/mdast-util-gfm-table/blob/main/from-markdown.js. So, also enter/exit as keys, which then are objects mapping token types to functions

File: package.json [hunk 1]
```diff
@@ -27,9 +27,12 @@
     "Titus Wormer <tituswormer@gmail.com> (https://wooorm.com)"
   ],
   "files": [
-    "index.js"
+    "index.js",
+    "index.d.ts"
   ],
+  "types": "index.d.ts",
   "dependencies": {
+    "@types/mdast": "^3.0.3",
```
Comment (wooorm) [inline package.json:35]:
```suggestion
    "@types/mdast": "^3.0.0",
```

**Review by wooorm — COMMENTED** [review]
one nit but otherwise good, @ChristianMurphy?

**Review by ChristianMurphy — COMMENTED** [review]
Thanks @ocavue!
Could we add a type tests for this to verify usage?
similar to https://github.com/micromark/micromark/blob/main/micromark.test.ts

**Comment by wooorm** [comment]
micromark 2.10 has types!

**Comment by ocavue** [comment]
haha. I was just doing some experiments between `export default` and `export =` and I saw the GitHub notification about the new version of micromark.

**Comment by wooorm** [comment]
Nice! ;)

And, can you apply that `export =` suggestion, or what did you find?

**Comment by ocavue** [comment]
Just added `export =`.

**Comment by wooorm** [comment]
Does it make sense to put all these files in a types/ directory? That's what we've been doing to most other stuff in unified

**Comment by ocavue** [comment]
Just moved them.

**Review by wooorm — APPROVED** [review]

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  mdastExtensions?: MdastExtension[]
+}
+
+declare function fromMarkdown(value: string | Buffer, options?: Options): Root
+
+declare function fromMarkdown(
+  value: string | Buffer,
+  encoding?: BufferEncoding,
+  options?: Options
+): Root
```
Comment (ChristianMurphy) [inline types/index.d.ts:31]: apologies I may have been unclear before, extra interfaces that we want exported when using `export =` need to be inside the namespace that matches the export.

```suggestion
declare namespace fromMarkdown {
  interface MdastExtension {
    enter: Record<Type, (token: Token) => void>
    exit: Record<Type, (token: Token) => void>
  }

  interface Options {
    extensions?: SyntaxExtension[]
    mdastExtensions?: MdastExtension[]
  }
}

declare function fromMarkdown(value: string | Buffer, options?: fromMarkdown.Options): Root

declare function fromMarkdown(
  value: string | Buffer,
  encoding?: BufferEncoding,
  options?: fromMarkdown.Options
): Root
```

File: types/mdast-util-from-markdown.test.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  formMarkdown(Buffer.alloc(8))
+
+  // $ExpectType Root
+  formMarkdown(Buffer.alloc(8), {extensions: []})
+
+  // $ExpectType Root
+  formMarkdown(Buffer.alloc(8), 'utf-8', {mdastExtensions: []})
+
+  // $ExpectError
+  formMarkdown(Buffer.alloc(8), 'utf-8', {allowDangerousHtml: true})
```
Comment (ChristianMurphy) [inline types/mdast-util-from-markdown.test.ts:22]:
```suggestion
import * as fromMarkdown from 'mdast-util-from-markdown'

function main() {
  const raw = '# text **strong**'

  // $ExpectType Root
  fromMarkdown(raw)

  // $ExpectType Root
  fromMarkdown(Buffer.alloc(8))

  // $ExpectType Root
  fromMarkdown(Buffer.alloc(8), {extensions: []})

  // $ExpectType Root
  fromMarkdown(Buffer.alloc(8), 'utf-8', {mdastExtensions: []})

  // $ExpectError
  fromMarkdown(Buffer.alloc(8), 'utf-8', {allowDangerousHtml: true})
```

**Review by ChristianMurphy — APPROVED** [review]

**Review by ChristianMurphy — APPROVED** [review]

**Review by wooorm — APPROVED** [review]

**Comment by wooorm** [comment]
Released, thanks @ocavue!

**Comment by ocavue** [comment]
You're welcome. It's an honor to be able to contribute to such an awesome project. @wooorm

**Comment by wooorm** [comment]
@ocavue are you also interested in doing to-markdown?

**Comment by ocavue** [comment]
I can do that. I should be able to submit a PR sometime this week. @wooorm

**Comment by wooorm** [comment]
@ocavue alight! I tried it myself (I didn't want to rush you but also want to publish remark soon): https://github.com/syntax-tree/mdast-util-to-markdown/pull/7

**Pull request merged**